### PR TITLE
1) Fixed bug im handling config items for agent settings

### PR DESCRIPTION
**Pull request by kalyan-nidumolu**
1) Fixed bug im handling config items for agent settings
2) Added additional unit test for config items.
3) Fixed debug log entry for reboot config counter
4) Updated docs for agent settings

Signed-off-by: Kalyan Nidumolu <kalyan@zadeda.com>

### CODE REVIEW

File: pkg/pillar/types/global_test.go [hunk 1]
```diff
@@ -197,63 +278,81 @@ func TestParseGlobalItem(t *testing.T) {
...
+		logLevel    AgentSettingKey
+		expectError bool
 	}{
-		"Agent Setting": {
+		"Agent Setting LogLevel New": {
 			logLevel: LogLevel,
 			item: config.ConfigItem{
 				Key:   "agent.zedagent.debug.loglevel",
 				Value: "info",
 			},
```
Comment (eriknordmark) [inline pkg/pillar/types/global_test.go:294]: Shouldn’t we test with something different than the default value?

File: docs/CONFIG-PROPERTIES.md [hunk 2]
```diff
@@ -30,9 +30,14 @@
...
+
+In addition, there can be per-agent settings.
+The Per-agent settings begin with "agent.*agentame*.*setting*"
+The following per-agent settings override the corresponding default ones:
 
 | Name | Type | Description |
 | ---- | ---- | ----------- |
-| debug.*agentname*.loglevel | string | if set overrides debug.default.loglevel |
-| debug.*agentname*.remote.loglevel | string | if set overrides debug.default.remote.loglevel |
+| debug.*agentname*.loglevel | string | if set overrides debug.default.loglevel | ( DEPRECATED by agent.*agentname*.loglevel)
```
Comment (eriknordmark) [inline docs/CONFIG-PROPERTIES.md:40]: Put the new ones here. Then parenthetically mention that the old ones are still supported

File: pkg/pillar/types/global_test.go [hunk 3]
```diff
@@ -126,6 +129,84 @@ func TestAddStringItem(t *testing.T) {
 		assert.Equal(t, test.expectedVal, specMap.GlobalSettings[test.key].StringDefault)
 	}
 }
+
+func TestNewConfigItemSpecMap(t *testing.T) {
+	// log.SetLevel(log.DebugLevel)
+	specMap := NewConfigItemSpecMap()
```
Comment (eriknordmark) [inline pkg/pillar/types/global_test.go:135]: So this is testing using the map we have in the product,right?

Comment (kalyan-nidumolu) [reply pkg/pillar/types/global_test.go:135]: Yes..

File: docs/CONFIG-PROPERTIES.md (hunk 2 above)
Comment (kalyan-nidumolu) [reply docs/CONFIG-PROPERTIES.md]: Good idea.. Done..

| Name | Type | Description |
| ---- | ---- | ----------- |
| agent.*agentname*.loglevel | string | if set overrides debug.default.loglevel | (Legacy setting debug.*agentname*.loglevel still supported)
| agent.*agentname*.remote.loglevel | string | if set overrides debug.default.remote.loglevel | (Legacy setting debug.*agentname*.remote.loglevel)

File: pkg/pillar/types/global_test.go (hunk 1 above)
Comment (kalyan-nidumolu) [reply pkg/pillar/types/global_test.go:376]: Good point.. Done..
Added more detailed test cases..

**Comment by kalyan-nidumolu** [comment]
> Thanks. Can we also have the test verify that the various settings an be set to non-default values? And not only for agent settings but also for the others.
> 
> If not that has to be tested in a real device.

I have one test case each for Int, String, TriState and  Bool, where we take a Global setting and assign some value..

**Comment by eriknordmark** [comment]
> I have one test case each for Int, String, TriState and Bool, where we take a Global setting and assign some value..

I'm not concerned about the parser being able to handle unit tests. I'm concerned about the actual set of config variables we specify in global.json are not specified correctly (as was the case for the missing agent settings.) How can we verify that all of those still exist and are of the correct type?

**Comment by kalyan-nidumolu** [comment]
> type

I added test case to verify none of the keys are missed. 
We currently don't have a test for the exact type for each key and their accepted values.. This would be duplicating all the data and was not sure.. But can be added ..
Let me go ahead and add it.. That means.. Any new config item needs to be added here as well.

**Pull request merged**